### PR TITLE
Autobahn run automation

### PR DESCRIPTION
**Pull request by derlih**
## What do these changes do?

automation of Autobahn tests

## Are there changes in behavior for the user?

no

## Related issue number

#4247 (automation only)

## Checklist

- [x] I think the code is well written
- [ ] Unit tests for the changes exist
- [ ] Documentation reflects the changes
- [x] If you provide code modification, please add yourself to `CONTRIBUTORS.txt`
  * The format is &lt;Name&gt; &lt;Surname&gt;.
  * Please keep alphabetical order, the file is sorted by names.
- [x] Add a new news fragment into the `CHANGES` folder
  * name it `<issue_id>.<type>` for example (588.bugfix)
  * if you don't have an `issue_id` change it to the pr id after creating the pr
  * ensure type is one of the following:
    * `.feature`: Signifying a new feature.
    * `.bugfix`: Signifying a bug fix.
    * `.doc`: Signifying a documentation improvement.
    * `.removal`: Signifying a deprecation or removal of public API.
    * `.misc`: A ticket has been closed, but it is not of interest to users.
  * Make sure to use full sentences with correct case and punctuation, for example: "Fix issue with non-ascii contents in doctest text files."

### CODE REVIEW

**Comment by derlih** [comment]
@asvetlov @webknjaz 
can you check this PR?

File: aiohttp/web_protocol.py [hunk 1]
```diff
@@ -3,6 +3,7 @@
 import traceback
 from collections import deque
 from contextlib import suppress
+from dataclasses import dataclass
```
Comment (webknjaz) [inline aiohttp/web_protocol.py:6]: How is this related to the change?

File: CHANGES/4247.misc [hunk 2]
```diff
@@ -0,0 +1 @@
+Autobahn test automation
```
Comment (webknjaz) [inline CHANGES/4247.misc:1]: Could you be more verbose and use the past tense?

File: tests/autobahn/Dockerfile.aiohttp [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+RUN \
+    apt-get update && \
+    apt-get -y install --no-install-recommends wait-for-it && \
+    apt-get clean && \
+    rm -rf /var/lib/apt/lists/*
+
+ADD ./ /src
+ENV AIOHTTP_NO_EXTENSIONS=1
+WORKDIR /src
+RUN python setup.py install
```
Comment (webknjaz) [inline tests/autobahn/Dockerfile.aiohttp:13]: <details>
<summary><strong>NOOOOOOOO!!111</strong></summary>

NOOOOOOOO!!111
</details>

Invoking `setup.py` directly is a very bad idea™. Pip should be used instead.

File: tests/autobahn/Dockerfile.aiohttp [hunk 4]
```diff
@@ -0,0 +1,13 @@
...
+
+ENV DEBIAN_FRONTEND=noninteractive
+RUN \
+    apt-get update && \
+    apt-get -y install --no-install-recommends wait-for-it && \
+    apt-get clean && \
+    rm -rf /var/lib/apt/lists/*
+
+ADD ./ /src
+ENV AIOHTTP_NO_EXTENSIONS=1
```
Comment (webknjaz) [inline tests/autobahn/Dockerfile.aiohttp:11]: Why?

File: tests/autobahn/server/server.py [hunk 5]
```diff
@@ -30,29 +29,16 @@ async def wshandler(request):
...
-    print("Server started at http://127.0.0.1:9001")
-    return app, srv, handler
-
-
-async def finish(app, srv, handler):
-    srv.close()
-    await handler.shutdown()
-    await srv.wait_closed()
+    web.run_app(app, port=9001)
+    print("Server started at http://0.0.0.0:9001")
```
Comment (webknjaz) [inline tests/autobahn/server/server.py:37]: This is misleading. `web.run_app()` blocks and `print()` is not executed after the server start but after the server shutdown.

File: tests/autobahn/Dockerfile.aiohttp [hunk 6]
```diff
@@ -0,0 +1,13 @@
+FROM python:3.9.1
+
+ENV DEBIAN_FRONTEND=noninteractive
+RUN \
+    apt-get update && \
+    apt-get -y install --no-install-recommends wait-for-it && \
```
Comment (webknjaz) [inline tests/autobahn/Dockerfile.aiohttp:6]: Why is this necessary? It looks like this is never used in this specific container.

File: tests/autobahn/images-for-compose.yml [hunk 7]
```diff
@@ -0,0 +1,13 @@
...
+    build:
+      context: .
+      dockerfile: Dockerfile.autobahn
+      args:
+        USER_ID: ${USER_ID}
+
+  aiohttp:
+    build:
+      context: ../..
+      dockerfile: tests/autobahn/Dockerfile.aiohttp
```
Comment (webknjaz) [inline tests/autobahn/images-for-compose.yml:13]: Building the base images when running tests is time-consuming. It's best to produce and push them to the GitHub Container Registry. You may draw some inspiration on doing this from https://github.com/ansible/pylibssh/blob/devel/.github/workflows/build-manylinux-container-images.yml.

File: tests/autobahn/run-tests.sh [hunk 8]
```diff
@@ -0,0 +1,12 @@
...
+rm -rf $PWD/reports
+mkdir $PWD/reports
+
+USER_ID=$(id -u) docker-compose -p aiohttp-autobahn -f $PWD/images-for-compose.yml build --parallel
+
+docker-compose -f $PWD/client/docker-compose.yml up --abort-on-container-exit
+docker-compose -f $PWD/client/docker-compose.yml down
+
+docker-compose -f $PWD/server/docker-compose.yml up --abort-on-container-exit
+docker-compose -f $PWD/server/docker-compose.yml down
```
Comment (webknjaz) [inline tests/autobahn/run-tests.sh:12]: Instead of having a shell script that never runs in CI, it's best to have it integrated into pytest (with something like https://pypi.org/p/pytest-docker-compose / https://pypi.org/p/lovely-pytest-docker / or similar) and GHA.

**Comment by webknjaz** [comment]
Hey, @derlih do you plan to resurrect this effort?

**Comment by webknjaz** [comment]
@paulefoe @anesabml do any of you want to pick up this PR?

**Comment by anesabml** [comment]
@webknjaz I'd love to as this is an opportunity for me to learn about Autobahn test suite, I will go over the changes in this PR and open a new one based on that.

**Comment by webknjaz** [comment]
Okay, fetch this branch, rebase and address the comments by adding new commits to your copy in a separate PR. Please also make sure to add the inline comments to the PR highlighting important places for the reviewers to notice + some explanation of the choices based on what I asked in this PR would be helpful too.

**Comment by derlih** [comment]
Hi
Unfortunately last month I didn't have much time for opensource :-(
I would be glad is someone continue the work of merging this automation part!
